test/e2e/dra/dra.go

```diff
@@ -775,6 +777,70 @@ var _ = ginkgo.Describe("[sig-node] DRA [Feature:DynamicResourceAllocation]", fu
 	})
 })
 
+// This test covers aspects of non graceful node shutdown by DRA controller
+// More details about this can be found in the KEP:
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown
+var _ = ginkgo.Describe("[sig-node] [Serial] [Disruptive] [Slow] [Feature:DynamicResourceAllocation] DRA: handling non graceful node shutdown", func() {
+	f := framework.NewDefaultFramework("dra")
+
+	// The driver containers have to run with sufficient privileges to
+	// modify /var/lib/kubelet/plugins.
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.Context("cluster", func() {
+		nodes := NewNodes(f, 2, 8)
+		driver := NewDriver(f, nodes, networkResources)
+		b := newBuilder(f, driver)
+
+		ginkgo.It("must deallocate on non graceful node shutdown", func(ctx context.Context) {
+			ginkgo.By("create test pod")
+			parameters := b.parameters()
+			label := "app.kubernetes.io/instance"
+			instance := f.UniqueName + "-test-app"
+			pod := b.podExternal()
+			pod.Labels[label] = instance
+			claim := b.externalClaim(resourcev1alpha2.AllocationModeWaitForFirstConsumer)
+			b.create(ctx, parameters, claim, pod)
+
+			ginkgo.By("wait for test pod " + pod.Name + " to run")
+			labelSelector := labels.SelectorFromSet(labels.Set(pod.Labels))
+			pods, err := e2epod.WaitForPodsWithLabelRunningReady(ctx, f.ClientSet, pod.Namespace, labelSelector, 1, framework.PodStartTimeout)
+			framework.ExpectNoError(err, "start pod")
+			runningPod := &pods.Items[0]
+
+			nodeName := runningPod.Spec.NodeName
+			// Prevent builder tearDown to fail waiting for unprepared resources
+			delete(b.driver.Nodes, nodeName)
+			ginkgo.By("stop node " + nodeName + " non gracefully")
+			ginkgo.DeferCleanup(framework.RunCmd, "docker", "start", nodeName)
+			_, stderr, err := framework.RunCmd("docker", "stop", nodeName)
+			gomega.Expect(stderr).To(gomega.BeEmpty())
+			framework.ExpectNoError(err)
+
+			ginkgo.By("apply out-of-service taint on node " + nodeName)
+			taint := v1.Taint{
+				Key:    v1.TaintNodeOutOfService,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			e2enode.AddOrUpdateTaintOnNode(ctx, f.ClientSet, nodeName, taint)
+			ginkgo.DeferCleanup(e2enode.RemoveTaintOffNode, f.ClientSet, nodeName, taint)
+
+			ginkgo.By("waiting for claim to get deallocated")
+			gomega.Eventually(ctx, func(ctx context.Context) (*resourcev1alpha2.ResourceClaim, error) {
+				return b.f.ClientSet.ResourceV1alpha2().ResourceClaims(b.f.Namespace.Name).Get(ctx, claim.Name, metav1.GetOptions{})
+			}).WithTimeout(f.Timeouts.PodDelete).Should(gomega.HaveField("Status.Allocation", (*resourcev1alpha2.AllocationResult)(nil)))
+
+			ginkgo.By("start node" + nodeName)
+			_, stderr, err = framework.RunCmd("docker", "start", nodeName)
+			gomega.Expect(stderr).To(gomega.BeEmpty())
+			framework.ExpectNoError(err)
```

Comment:
You can move all of this into the `DeferCleanup` above. A cleanup callback can fail just like the `It` callback and it will make the test as failed.

Reply:
forget about it. It seems to work correctly without it. I'll remove this.

Reply:
Without this test takes much longer. I've commented in the PR description about it. Now I've added a comment to the test code.

Reply:
I wasn't suggesting to remove `framework.RunCmd("docker", "start", nodeName)`, just to put it into a `ginkgo.DeferCleanup`. Why is that slower? It shouldn't be.

Reply:
`ginkgo.DeferCleanup(framework.RunCmd, "docker", "start", nodeName)` is in the code from the very beginning.
However, without explicitly calling `docker start` at the end of the test, the test takes a lot more time.